### PR TITLE
[Flang][OpenMP] Add some semantic checks for Linear clause

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
This PR adds all the missing semantics for the Linear clause based on the OpenMP 5.2 restrictions. The restriction details are mentioned below.

OpenMP 5.2:
5.4.6 linear Clause restrictions
- A linear-modifier may be specified as ref or uval only on a declare simd directive.
- If linear-modifier is not ref, all list items must be of type integer.
- If linear-modifier is ref or uval, all list items must be dummy arguments without the VALUE attribute.
- List items must not be Cray pointers or variables that have the POINTER attribute. Cray pointer support has been deprecated.
- If linear-modifier is ref, list items must be polymorphic variables, assumed-shape arrays, or variables with the ALLOCATABLE attribute. 
- A common block name must not appear in a linear clause.
- The list-item cannot appear more than once

4.4.4 ordered Clause restriction
- If n is explicitly specified, a linear clause must not be specified on the same directive.

5.11 aligned Clause restriction
- Each list item must have C_PTR or Cray pointer type or have the POINTER or ALLOCATABLE attribute. Cray pointer support has been deprecated.

### CODE REVIEW

File: flang/test/Semantics/OpenMP/linear-cluase01.f90
Comment (kparzysz) [inline flang/test/Semantics/OpenMP/linear-cluase01.f90:1]: Typo: cluase -> clause.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 1]
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
 void OmpStructureChecker::Enter(const parser::OmpClause::Linear &x) {
   CheckAllowedClause(llvm::omp::Clause::OMPC_linear);
 
+  parser::CharBlock source = GetContext().clauseSource;
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:3003]: Use braced initialized in frontend code.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 2]
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+  // 4.4.4 Ordered clause restriction
+  if (const auto *clause{
+          FindClause(GetContext(), llvm::omp::Clause::OMPC_ordered)}) {
+    const auto &orderedClause{std::get<parser::OmpClause::Ordered>(clause->u)};
+    if (orderedClause.v) {
+      return;
     }
   }
+
+  auto checkForValidLinearClaue = [&](const parser::Name &name, bool is_ref) {
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:3025]:
```suggestion
  auto checkForValidLinearClause = [&](const parser::Name &name, bool is_ref) {
```

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 3]
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+    parser::CharBlock source = GetContext().clauseSource;
+    std::string list_item_name = name.ToString();
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:3027]:
```suggestion
    std::string listItemName = name.ToString();
```

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 4]
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+  if (const auto *clause{
+          FindClause(GetContext(), llvm::omp::Clause::OMPC_ordered)}) {
+    const auto &orderedClause{std::get<parser::OmpClause::Ordered>(clause->u)};
+    if (orderedClause.v) {
+      return;
     }
   }
+
+  auto checkForValidLinearClaue = [&](const parser::Name &name, bool is_ref) {
+    parser::CharBlock source = GetContext().clauseSource;
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:3026]: Nit: Use braced initialization.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 5]
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+              }
+            }
+          },
+          [&](const parser::OmpLinearClause::WithModifier &withModifier) {
+            for (const auto &name : withModifier.names) {
+              if (name.symbol) {
+                checkForValidLinearClaue(name,
+                    (withModifier.modifier.v ==
+                        parser::OmpLinearModifier::Type::Ref));
+                std::string list_item_name = name.ToString();
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:3071]:
```suggestion
                std::string listItemName = name.ToString();
```

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 6]
```diff
@@ -303,6 +303,30 @@ void OmpStructureChecker::CheckMultListItems() {
     CheckMultipleOccurrence(
         listVars, nontempNameList, itr->second->source, "NONTEMPORAL");
   }
+
+  // Linear clause
+  auto linearClauses{FindClauses(llvm::omp::Clause::OMPC_linear)};
+  for (auto itr = linearClauses.first; itr != linearClauses.second; ++itr) {
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:309]:
```suggestion
  for (auto itr{linearClauses.first}; itr != linearClauses.second; ++itr) {
```

**Review by kiranchandramohan — COMMENTED** [review]
A few nit comments.

Could you expand the summary to list the checks implemented in the patch with a reference to the standard and section?

File: flang/test/Semantics/OpenMP/linear-cluase01.f90
Comment (Thirumalai-Shaktivel) [reply flang/test/Semantics/OpenMP/linear-cluase01.f90:1]: Done.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 7]
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+      return;
+    }
+  }
+
+  // OpenMP 5.2: Ordered clause restriction
+  if (const auto *clause{
+          FindClause(GetContext(), llvm::omp::Clause::OMPC_ordered)}) {
+    const auto &orderedClause{std::get<parser::OmpClause::Ordered>(clause->u)};
+    if (orderedClause.v) {
+      return;
```
Comment (Thirumalai-Shaktivel) [inline flang/lib/Semantics/check-omp-structure.cpp:3021]: Flang-new already throws an error, `Clause LINEAR is not allowed if clause ORDERED appears on the DO directive`. So, I skipped here.

**Comment by Thirumalai-Shaktivel** [comment]
I addressed all the comments, @kiranchandramohan.

Thanks for the reviews.

File: flang/test/Semantics/OpenMP/declarative-directive.f90 [hunk 8]
```diff
@@ -23,6 +23,7 @@ end subroutine requires_2
 
 subroutine declare_simd_1(a, b)
   real(8), intent(inout) :: a, b
+  !ERROR: 'a' in ALIGNED clause must be of type C_PTR, POINTER or ALLOCATABLE
```
Comment (NimishMishra) [inline flang/test/Semantics/OpenMP/declarative-directive.f90:26]: Do we also deal with ALIGNED clause in this PR? A similar issue is causing the Linux CI failure:

```
flang/test/Examples/omp-declarative-directive.f90:11:38: error: 'a' in ALIGNED clause must be of type C_PTR, POINTER or ALLOCATABLE
--
  | !$omp declare simd(declare_simd_1) aligned(a)
```

**Review by NimishMishra — COMMENTED** [review]
One small comment on ALIGNED clause test added in this PR.

Ultra nit: "If n is explicitly specified, a linear clause must not be specified on the same directive" in description -> "If an ordered clause is explicitly specified......"

File: flang/test/Semantics/OpenMP/declarative-directive.f90 (hunk 8 above)
Comment (Thirumalai-Shaktivel) [reply flang/test/Semantics/OpenMP/declarative-directive.f90:26]: Yes, this line: https://github.com/llvm/llvm-project/pull/111354/files#diff-be225bc038e95a229024c9f227284112f98fe31762bc0bb80605de425aadd084R2287 is responsible for the change.

Comment (Thirumalai-Shaktivel) [reply flang/test/Semantics/OpenMP/declarative-directive.f90:26]: I fixed it in the recent commit.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 9]
```diff
@@ -323,6 +323,30 @@ void OmpStructureChecker::CheckMultListItems() {
     CheckMultipleOccurrence(
         listVars, nontempNameList, itr->second->source, "NONTEMPORAL");
   }
+
+  // Linear clause
+  auto linearClauses{FindClauses(llvm::omp::Clause::OMPC_linear)};
+  for (auto itr{linearClauses.first}; itr != linearClauses.second; ++itr) {
+    const auto &linearClause{
+        std::get<parser::OmpClause::Linear>(itr->second->u)};
```
Comment (kparzysz) [inline flang/lib/Semantics/check-omp-structure.cpp:331]: `FindClauses` now returns an iterator range, please use range-for here.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 10]
```diff
@@ -323,6 +323,30 @@ void OmpStructureChecker::CheckMultListItems() {
...
+              for (const auto &name : withoutModifier.names) {
+                nameList.push_back(name);
+              }
+            },
+            [&](const parser::OmpLinearClause::WithModifier &withModifier) {
+              for (const auto &name : withModifier.names) {
+                nameList.push_back(name);
+              }
+            },
+        },
```
Comment (kparzysz) [inline flang/lib/Semantics/check-omp-structure.cpp:346]: You can only have a single visitor, plus you can simplify the copy:
```
[&](auto &&s) {
  std::copy(s.names.begin(), s.names.end(), std::back_inserter(nameList));
}
```

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 11]
```diff
@@ -3040,16 +3064,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+          [&](const parser::OmpLinearClause::WithoutModifier &withoutModifier) {
+            for (const auto &name : withoutModifier.names) {
+              if (name.symbol) {
+                checkForValidLinearClause(name, false);
+              }
+            }
+          },
+          [&](const parser::OmpLinearClause::WithModifier &withModifier) {
+            for (const auto &name : withModifier.names) {
+              if (name.symbol) {
```
Comment (kparzysz) [inline flang/lib/Semantics/check-omp-structure.cpp:3131]: You can extract this into a lamda to reduce indentation.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 12]
```diff
@@ -3040,16 +3064,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+                    (withModifier.modifier.v ==
+                        parser::OmpLinearModifier::Type::Ref));
+                std::string listItemName{name.ToString()};
+                if (withModifier.modifier.v !=
+                        parser::OmpLinearModifier::Type::Val &&
+                    IsDummy(*name.symbol) && IsValue(*name.symbol)) {
+                  context_.Say(source,
+                      "The list item `%s` specified with the linear-modifier "
+                      "`REF` or `UVAL` must be a dummy argument without "
+                      "`VALUE` attribute"_err_en_US,
```
Comment (kparzysz) [inline flang/lib/Semantics/check-omp-structure.cpp:3142]: Please concatenate messages into a single string.  This is to make grepping for messages easier.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 13]
```diff
@@ -3040,16 +3064,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+                if (withModifier.modifier.v ==
+                        parser::OmpLinearModifier::Type::Ref &&
+                    !(IsAllocatable(*name.symbol) ||
+                        IsAssumedShape(*name.symbol) ||
+                        IsPolymorphic(*name.symbol))) {
+                  context_.Say(source,
+                      "The list item `%s` specified with the linear-modifier "
+                      "`REF` must be polymorphic variable, assumed-shape "
+                      "array, "
+                      "or a variable with the `ALLOCATABLE` attribute"_err_en_US,
```
Comment (kparzysz) [inline flang/lib/Semantics/check-omp-structure.cpp:3154]: Same here.

**Comment by kparzysz** [comment]
I will change the modifier representation in LINEAR in a week or so.  Please finish this soon, otherwise my code will create conflicts for you.

**Comment by Thirumalai-Shaktivel** [comment]
Thanks for the review, @kparzysz!

I addressed all the comments. If anything else to be done, please let me know.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 14]
```diff
@@ -3542,16 +3557,98 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+        }
+      };
+
+  // OpenMP 5.2: Linear clause Restrictions
+  common::visit(
+      common::visitors{
+          [&](const parser::OmpLinearClause::WithoutModifier &withoutModifier) {
+            for (const auto &name : withoutModifier.names)
+              if (name.symbol)
+                checkForValidLinearClause_01(name, false);
```
Comment (kparzysz) [inline flang/lib/Semantics/check-omp-structure.cpp:3642]: Please add {} for nested statements.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 15]
```diff
@@ -383,6 +383,21 @@ void OmpStructureChecker::CheckMultListItems() {
...
+  for (auto [_, clause] : FindClauses(llvm::omp::Clause::OMPC_linear)) {
+    const auto &linearClause{std::get<parser::OmpClause::Linear>(clause->u)};
+    std::list<parser::Name> nameList;
+    common::visit(
+        common::visitors{
+            [&](const auto &u) {
+              std::copy(
+                  u.names.begin(), u.names.end(), std::back_inserter(nameList));
+            },
+        },
```
Comment (kparzysz) [inline flang/lib/Semantics/check-omp-structure.cpp:397]: You only have one visitor, so you don't need `common::visitors`.  Do `common::visit([&]...)`.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 16]
```diff
@@ -3542,16 +3557,98 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+  if (const auto *clause{
+          FindClause(GetContext(), llvm::omp::Clause::OMPC_ordered)}) {
+    const auto &orderedClause{std::get<parser::OmpClause::Ordered>(clause->u)};
+    if (orderedClause.v) {
+      return;
     }
   }
+
+  auto checkForValidLinearClause_01 = [&](const parser::Name &name,
+                                          bool is_ref) {
```
Comment (kparzysz) [inline flang/lib/Semantics/check-omp-structure.cpp:3583]: Please concatenate all error message strings in this function.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 17]
```diff
@@ -3542,16 +3557,98 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
...
+      context_.Say(source,
+          "'%s' is a common block name and must not appear in an "
+          "LINEAR clause"_err_en_US,
+          listItemName);
+    }
+  };
+
+  auto checkForValidLinearClause_02 =
+      [&](const parser::Name &name,
+          const parser::OmpLinearModifier::Value &modifierValue) {
```
Comment (kparzysz) [inline flang/lib/Semantics/check-omp-structure.cpp]: Please concatenate all error message strings in this function.

**Review by kparzysz — COMMENTED** [review]
Almost ready, a few more comments.

File: flang/lib/Semantics/check-omp-structure.cpp (hunk 14 above)
Comment (Thirumalai-Shaktivel) [reply flang/lib/Semantics/check-omp-structure.cpp:3642]: Do you mean the following?
```suggestion
            for (const auto &name : withoutModifier.names) {
              if (name.symbol) {
                checkForValidLinearClause_01(name, false);
              }
            }
```

File: flang/lib/Semantics/check-omp-structure.cpp (hunk 15 above)
Comment (Thirumalai-Shaktivel) [reply flang/lib/Semantics/check-omp-structure.cpp:397]: Yea, sure.

File: flang/lib/Semantics/check-omp-structure.cpp (hunk 16 above)
Comment (Thirumalai-Shaktivel) [reply flang/lib/Semantics/check-omp-structure.cpp:3583]: I'm sorry, but I'm not getting this.

Concatenate in the sense, for the following:
```
context_.Say(source,
          "The list item `%s` specified with other than linear-modifier `REF`"
          " must be of type INTEGER"_err_en_US,
          listItemName);
```
should I do?
```
context_.Say(source,
          "The list item `%s` specified with other than linear-modifier `REF` must be of type INTEGER"_err_en_US,
          listItemName);
```

Comment (kparzysz) [reply flang/lib/Semantics/check-omp-structure.cpp:3583]: Yes

File: flang/lib/Semantics/check-omp-structure.cpp (hunk 14 above)
Comment (kparzysz) [reply flang/lib/Semantics/check-omp-structure.cpp:3642]: Yes

**Comment by kparzysz** [comment]
Please don't use force-pushes.  That messes up previous comments.

**Comment by Thirumalai-Shaktivel** [comment]
Sorry about that.

I rebased the commits and I had to force-push the changes.
Next time I will make sure not to rebase it.

**Comment by Thirumalai-Shaktivel** [comment]
Done.

**Review by kparzysz — APPROVED** [review]

**Comment by Thirumalai-Shaktivel** [comment]
Thank you!

**Pull request merged**